File: apps/android/app/src/main/java/com/karibuhealth/app/domain/AncProtocol.kt
package com.karibuhealth.app.domain

import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * ANC protocol scheduling + gap detection per Uganda MOH (docs/maternal-neonatal-research.md).
 * Pure and deterministic so the registry can flag "due / overdue / behind" without
 * the server. `today` is injectable for testing.
 *
 * Uganda adopted the WHO 8-contact model (ANC8). IPTp-SP target is >=3 doses from
 * the 2nd trimester. EDD = LMP + 280 days (Naegele).
 */
object AncProtocol {

    /** ANC8 recommended contact weeks (first by 12). */
    val SCHEDULE_WEEKS = listOf(12, 20, 26, 30, 34, 36, 38, 40)
    const val IPTP_TARGET = 3
    const val TERM_WEEKS = 40
    private const val IPTP_START_WEEK = 20  // 2nd trimester onward

    fun eddFromLmp(lmp: LocalDate): LocalDate = lmp.plusDays(280)

    fun gestationWeeks(lmp: LocalDate?, today: LocalDate = LocalDate.now()): Int? =
        lmp?.let { ChronoUnit.WEEKS.between(it, today).toInt().coerceAtLeast(0) }

    /** How many ANC8 contacts should have happened by this gestation. */
    fun contactsDue(gestationWeeks: Int?): Int =
        if (gestationWeeks == null) 0 else SCHEDULE_WEEKS.count { it <= gestationWeeks }

    data class Status(
        val gestationWeeks: Int?,
        val contactsDone: Int,
        val contactsDue: Int,
        val iptpDone: Int,
        val ancBehind: Boolean,
        val iptpBehind: Boolean,
        val dueSoon: Boolean,    // EDD within 14 days
        val postDates: Boolean,  // past EDD / >42 weeks
        /** Short human-readable gap chips for the registry. */
        val gaps: List<String>,
    )

    fun status(
        lmp: LocalDate?,
        edd: LocalDate?,
        contactsDone: Int,
        iptpDone: Int,
        today: LocalDate = LocalDate.now(),
    ): Status {
        val ga = gestationWeeks(lmp, today)
        val due = contactsDue(ga)
        val ancBehind = contactsDone < due
        // Behind on IPTp once the 2nd-trimester window is open and target not met.
        val iptpBehind = (ga != null && ga >= IPTP_START_WEEK && iptpDone < IPTP_TARGET)
        val effectiveEdd = edd ?: lmp?.let { eddFromLmp(it) }
        val postDates = effectiveEdd != null && today.isAfter(effectiveEdd)
        val dueSoon = effectiveEdd != null && !postDates &&
            !today.isAfter(effectiveEdd) &&
            ChronoUnit.DAYS.between(today, effectiveEdd) <= 14

        val gaps = buildList {
            if (ancBehind) add("ANC $contactsDone/$due")
            if (iptpBehind) add("IPTp $iptpDone/$IPTP_TARGET")
            if (postDates) add("Post-dates")
            else if (dueSoon) add("Due soon")
        }
        return Status(ga, contactsDone, due, iptpDone, ancBehind, iptpBehind, dueSoon, postDates, gaps)
    }
}

File: apps/web/src/lib/anc-protocol.ts
/**
 * ANC protocol scheduling + gap detection (Uganda MOH ANC8).
 * Mirrors apps/android/.../domain/AncProtocol.kt
 */

export const ANC_SCHEDULE_WEEKS = [12, 20, 26, 30, 34, 36, 38, 40] as const
export const IPTP_TARGET = 3
const IPTP_START_WEEK = 20

export type AncProtocolStatus = {
  gestationWeeks: number | null
  contactsDone: number
  contactsDue: number
  iptpDone: number
  ancBehind: boolean
  iptpBehind: boolean
  dueSoon: boolean
  postDates: boolean
  gaps: string[]
}

function parseDate(iso: string | null | undefined): Date | null {
  if (!iso) return null
  const d = new Date(iso.slice(0, 10))
  return Number.isNaN(d.getTime()) ? null : d
}

function daysBetween(a: Date, b: Date): number {
  const ms = b.getTime() - a.getTime()
  return Math.floor(ms / (24 * 60 * 60 * 1000))
}

function weeksBetween(start: Date, end: Date): number {
  return Math.max(0, Math.floor(daysBetween(start, end) / 7))
}

export function eddFromLmp(lmp: Date): Date {
  const d = new Date(lmp)
  d.setDate(d.getDate() + 280)
  return d
}

export function gestationWeeks(lmp: string | null | undefined, today = new Date()): number | null {
  const lmpDate = parseDate(lmp)
  if (!lmpDate) return null
  return weeksBetween(lmpDate, today)
}

export function contactsDue(gestationWeeksVal: number | null): number {
  if (gestationWeeksVal == null) return 0
  return ANC_SCHEDULE_WEEKS.filter((w) => w <= gestationWeeksVal).length
}

export function ancProtocolStatus(input: {
  lmp: string | null | undefined
  edd: string | null | undefined
  contactsDone: number
  iptpDone: number
  today?: Date
}): AncProtocolStatus {
  const today = input.today ?? new Date()
  const ga = gestationWeeks(input.lmp, today)
  const due = contactsDue(ga)
  const ancBehind = input.contactsDone < due
  const iptpBehind = ga != null && ga >= IPTP_START_WEEK && input.iptpDone < IPTP_TARGET

  const lmpDate = parseDate(input.lmp)
  const eddDate = parseDate(input.edd) ?? (lmpDate ? eddFromLmp(lmpDate) : null)

  const postDates = eddDate != null && today > eddDate
  const dueSoon =
    eddDate != null &&
    !postDates &&
    today <= eddDate &&
    daysBetween(today, eddDate) <= 14

  const gaps: string[] = []
  if (ancBehind) gaps.push(`ANC ${input.contactsDone}/${due}`)
  if (iptpBehind) gaps.push(`IPTp ${input.iptpDone}/${IPTP_TARGET}`)
  if (postDates) gaps.push('Post-dates')
  else if (dueSoon) gaps.push('Due soon')

  return {
    gestationWeeks: ga,
    contactsDone: input.contactsDone,
    contactsDue: due,
    iptpDone: input.iptpDone,
    ancBehind,
    iptpBehind,
    dueSoon,
    postDates,
    gaps,
  }
}

/** Derive LMP from gestation weeks entered at registration (matches Android). */
export function lmpFromGestationWeeks(weeks: number, today = new Date()): string {
  const d = new Date(today)
  d.setDate(d.getDate() - weeks * 7)
  return d.toISOString().slice(0, 10)
}
